platform/security/idp/security-idp-plugin/security-idp-plugin-namequalifier/src/main/java/org/codice/ddf/security/idp/plugin/namequalifier/NameQualifierPresignPlugin.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.idp.plugin.namequalifier;
+
+import ddf.security.samlp.SamlProtocol;
+import java.util.List;
+import java.util.Set;
+import org.codice.ddf.configuration.SystemBaseUrl;
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.Response;
+
+/**
+ * Pre-sign plugin that ensures that "Each bearer assertion must contain an {@code <nameQualifier>}"
+ * as per section 8.3.7 of the SAML Spec.
+ *
+ * @see <a
+ *     href="https://www.oasis-open.org/committees/download.php/56783/sstc-saml-profiles-errata-2.0-wd-07-diff.pdf">
+ *     Profiles for the OASIS Security Assertion Markup Language (SAML) V2.0 – Errata Composite</a>
+ */
+public class NameQualifierPresignPlugin implements SamlPresignPlugin {
+
+  private static final String NAME_QUALIFIER =
+      String.format(
+          "https://%s:%s/services/idp/login",
+          SystemBaseUrl.INTERNAL.getHost(), SystemBaseUrl.INTERNAL.getHttpsPort());
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<SamlProtocol.Binding> supportedBindings) {
+
+    response.getIssuer().setNameQualifier(NAME_QUALIFIER);
```

Comment:
I think this is specifically for Persistent IDs so you might have to check that the format is `urn:oasis:names:tc:SAML:2.0:nameid-format:persistent`.

Reply:
I checked on the debugger and the format in the response is always null, so I'm not sure where to check. @brjeter thoughts?

Reply:
Yes sorry that wasn't clear. before setting the NameQualifier, you should check that the `Format` attribute on both the `<Issuer>` and the `<NameID>`s is equal to `urn:oasis:names:tc:SAML:2.0:nameid-format:persistent` because according to the spec, it only applies to persistent identifiers.

Reply:
❗️ Side Note: If we get to this plugin and the NameQualifier is already set, it should not be overridden. In other words, only set the issuer's NameQualifier if it is empty/null.

Reply:
@bakejeyner not sure what you mean by that. I thought the `NameQualifier` will be set but the value is incorrect which is why we have this plugin to set it to our IdP entity ID.